DrawerViewExample/DrawerViewExample/DrawerView.swift

```diff
@@ -0,0 +1,281 @@
+    case .changed:
+      animateCollapsing(fromGesture: gestureRecogniser)
+    case .ended:
+      finishCollapsingAnimation(fromGesture: gestureRecogniser)
+    default:
+      break
+    }
+  }
+
+  private func startCollapsingAnimation(fromGesture gestureRecogniser: UIPanGestureRecognizer) {
+    guard let superView = superview else {
+      return
+    }
+    superView.setNeedsLayout()
+    superView.layoutIfNeeded()
+    initialTouchPosition = gestureRecogniser.translation(in: superView).y
+
+    // filter view will collapse
+    propertyAnimator = animatorForCollapseGesture()
+    propertyAnimator.startAnimation()
+    propertyAnimator.pauseAnimation()
+  }
+
+  private func animatorForCollapseGesture() -> UIViewPropertyAnimator {
+    let animator = UIViewPropertyAnimator(duration: .animationDuration, curve: .easeIn, animations: { [weak self] in
+      self?.animatingConstraint.constant = .initialOffset
+      self?.superview?.layoutIfNeeded()
+    })
+    animator.addCompletion { [weak self] position in
+      self?.settleCollapsingAnimation(fromPosition: position)
+    }
+    return animator
+  }
+
+  private func animateCollapsing(fromGesture gestureRecogniser: UIPanGestureRecognizer) {
+    guard let superView = superview else {
+      return
+    }
+
+    let fullLength: CGFloat = .initialOffset - upperAnimationLimit
+    let currentFraction = (gestureRecogniser.translation(in: superView).y - initialTouchPosition)
+      .asProportion(of: fullLength)
+      .clamped(between: 0.01, and: 0.99)
+    propertyAnimator.fractionComplete = currentFraction
+  }
+
+  private func finishCollapsingAnimation(fromGesture gestureRecogniser: UIPanGestureRecognizer) {
+    guard let superView = superview else {
+      return
+    }
+
+    let fullLength: CGFloat = .initialOffset - upperAnimationLimit
+    let currentFraction = (initialTouchPosition - gestureRecogniser.translation(in: superView).y) / fullLength
+
+    if abs(currentFraction) < .tippingPoint {
+      propertyAnimator.isReversed = true
+    }
+
+    propertyAnimator.startAnimation()
+  }
+
+  private func settleCollapsingAnimation(fromPosition position: UIViewAnimatingPosition) {
+    if position == .start {
+      // this is called when trying to collapse the filters but letting them snap up again
+      delegate?.filterViewWillExpand(self)
+      animatingConstraint.constant = upperAnimationLimit
+      superview?.layoutIfNeeded()
+    } else if position == .end {
+      delegate?.filterViewDidCollapse(self)
+      panToExpandGestureRecogniser.isEnabled = true
+      panToCollapseGestureRecogniser.isEnabled = false
+    }
+    superview?.layoutIfNeeded()
+  }
+}
+
+private extension TimeInterval {
+  static let animationDuration: TimeInterval = 0.2
+}
+
+private extension CGFloat {
+    
+    func asProportion(of value: CGFloat) -> CGFloat {
+        return self / value
+    }
+    
+    func clamped(between minimum: CGFloat, and maximum: CGFloat) -> CGFloat {
```

Comment:
can be extracted to
```
extension Comparable {
    func clamped(to limits: ClosedRange<Self>) -> Self {
        return min(max(self, limits.lowerBound), limits.upperBound)
    }
}
```

Reply:
done! 
https://github.com/novoda/spikes/pull/398/commits/2afc38e4de9b85ea1fa56dd10de6061466b20c1e